#5: Allow ember-intl Translations

srsgores opened the pull request:
Not entirely tested, as tree fails (even in current master) with the following:

> Promise rejected before it renders with the right props: jQuery(...).on is not a function@ 155 ms
> Source:  
> TypeError: jQuery(...).on is not a function
>     at _setupAJAXHooks (http://localhost:4200/assets/test-support.js:7851:22)
>     at Klass.setupAJAXListeners (http://localhost:4200/assets/test-support.js:6778:49)
>     at http://localhost:4200/assets/test-support.js:6732:26
>     at Object.initializePromise (http://localhost:4200/assets/vendor.js:63961:7)
>     at new Promise (http://localhost:4200/assets/vendor.js:65813:21)
>     at nextStep (http://localhost:4200/assets/test-support.js:6731:18)
>     at tryCatch (http://localhost:4200/assets/vendor.js:63911:14)
>     at Object.invokeCallback (http://localhost:4200/assets/vendor.js:63926:15)
>     at http://localhost:4200/assets/vendor.js:66054:25
>     at http://localhost:4200/assets/vendor.js:42159:7
> Died on test #2     at testWrapper (http://localhost:4200/assets/test-support.js:6645:11)
>     at test (http://localhost:4200/assets/test-support.js:6658:39)
>     at Module.callback (http://localhost:4200/assets/tests.js:164:24)
>     at Module.exports (http://localhost:4200/assets/vendor.js:93:34)
>     at Module.build (http://localhost:4200/assets/vendor.js:145:10)
>     at findModule (http://localhost:4200/assets/vendor.js:196:9)
>     at requireModule (http://localhost:4200/assets/vendor.js:184:12): this.subject is not a function@ 199 ms
> Source:  
> TypeError: this.subject is not a function
>     at Object.<anonymous> (http://localhost:4200/assets/tests.js:166:26)
>     at Object.wrapper (http://localhost:4200/assets/test-support.js:6616:29)
>     at runTest (http://localhost:4200/assets/test-support.js:2796:28)
>     at Object.run (http://localhost:4200/assets/test-support.js:2781:4)
>     at http://localhost:4200/assets/test-support.js:2923:11
>     at process (http://localhost:4200/assets/test-support.js:2582:24)
>     at begin (http://localhost:4200/assets/test-support.js:2564:2)
>     at http://localhost:4200/assets/test-support.js:2624:4
> Promise rejected after it renders with the right props: Cannot read property 'subject' of undefined@ 212 ms
> Source:  
> TypeError: Cannot read property 'subject' of undefined
>     at Klass.teardownSubject (http://localhost:4200/assets/test-support.js:7669:31)
>     at http://localhost:4200/assets/test-support.js:6732:26
>     at Object.initializePromise (http://localhost:4200/assets/vendor.js:63961:7)
>     at new Promise (http://localhost:4200/assets/vendor.js:65813:21)
>     at nextStep (http://localhost:4200/assets/test-support.js:6731:18)
>     at Klass.invokeSteps (http://localhost:4200/assets/test-support.js:6738:14)
>     at http://localhost:4200/assets/test-support.js:6698:23
>     at tryCatch (http://localhost:4200/assets/vendor.js:63911:14)
>     at Object.invokeCallback (http://localhost:4200/assets/vendor.js:63926:15)
>     at http://localhost:4200/assets/vendor.js:66054:25

srsgores committed:
#5: Add translation service
Allow multiple translation services such as ember-i18n and ember-intl
Add unit tests for translation service

srsgores committed:
#5: Add translation service
Allow multiple translation services such as ember-i18n and ember-intl
Add unit tests for translation service

srsgores committed:
#5: use new translation service in inputs and labels

srsgores committed:
#5: use new translation service in inputs and labels

piotrekbator commented on addon/components/simple-input.js:
```diff
@@ -3,7 +3,7 @@ import humanize from "ember-form-builder/utilities/humanize";
 
 const extension = {
-  i18n: Ember.inject.service(),
+  translationService: Ember.inject.service(),
```
How about naming this service `formBuilderTranslations` instead of `translationService`?

srsgores replied:
`formBuilderTranslations` suggests that there are **multiple** translations, when it's just an instance of the preferred translation service.  That sounds misleading.

piotrekbator replied:
So how about `formBuilderI18n` or `formBuilderTranslator`?

piotrekbator commented on addon/components/simple-input.js:
```diff
@@ -141,7 +141,7 @@ const extension = {
 
     var result;
-    if (this.get("i18n") && this.get("i18n").exists(key)) { result = this.get("i18n").t(key); }
+    if (this.get("translationService.hasTranslationService") && this.get("translationService.translationService").exists(key)) { result = this.get("translationService.translationService").t(key); }
```
I'd prefer if the service encapsulated the `exists` and `t` method. In this case we depend in whole addon on the fact that both `ember-i18n` and `ember-intl` have those methods. I think it'd be much cleaner and safer if we abstracted this knowledge.

As long as both those libraries have both those methods simple delegation would be enough: `exists() { return this.get("translationService").exists(...arguments); }`.

piotrekbator commented on addon/services/translation.js:
```diff
@@ -0,0 +1,12 @@
+  translationServices: Ember.computed.collect("i18n", "intl"),
+  hasTranslationService: Ember.computed.notEmpty("translationServices"),
+  activeTranslationServices: Ember.computed("translationServices", function() {
+    return this.get("hasTranslationService") ? Ember.A() : this.get("translationServices").compact();
```
Shouldn't this be exactly opposite? An empty array when there are is no translation service?

piotrekbator commented on tests/unit/services/translation-test.js:
```diff
@@ -0,0 +1,29 @@
+    i18n: null
+  });
+  // TODO: assert using sinon: http://sinonjs.org/qunit/
+  assert.ok(false);
```
I don't think we need this warning -- a translation library is not required for `ember-form-builder` to work -- you can always depend on automatically generated labels or provide your own via the `label` attribute.

piotrekbator commented:
Thank's a lot. The PR looks very very nice! :)

I've made a few comments, please let me know what you think.

As for the error -- I'm not sure what that might be, I'll try to recreate this on my machine tomorrow.

piotrekbator commented:
BTW I can see in Travis that most of the tests regarging translations are failing. I used to inject fake i18n-like object to formBuilder in those tests -- now we should stub translations in the new service. Do you think you'll be able to do this?

piotrekbator commented:
@srsgores I've tried repeating the problem you're encountering but without luck. Can you tell me which version of node are you on? I have this working on 5.9 locally and it passess tests (on master) on Travis using 0.12. No problem similar to yours in any of those cases :|

srsgores committed:
#5: Update tests to check defaulting to i18n
Fix empty array always returning in activeTranslationService in ternary expression
Delegate t() and exists() methods to translation service

srsgores committed:
#5: Update tests to check defaulting to i18n
Fix empty array always returning in activeTranslationService in ternary expression
Delegate t() and exists() methods to translation service

srsgores commented:
@piotrekbator, thanks for your feedback.  I've adjusted the pull request to fix the wrong ternary expression, and delegated `t()` and `exists()` to the translation service.

Tests are still failing on `node 5.10.0` (windows).

srsgores commented:
Has there been any update on this?  @piotrekbator, are all tests failing on your end?

piotrekbator commented:
Sorry, but I'm unable to reproduce the problems you're experiencing. The tests run correctly, although some of them fail. Travis is also able to process the tests, you can check it out here: https://travis-ci.org/nibynic/ember-form-builder/builds/129567600

I don't have access to a Windows machine that I could use to install node etc. and I also have no idea why Windows installation would produce the results you're experiencing. Do you think you could try running it on any unix system?

srsgores committed:
#6: Fix test failing on translation service

srsgores committed:
#6: Fix test failing on translation service

srsgores committed:
#5: Fix translation service not found

srsgores committed:
#5: Fix translation service not found

piotrekbator committed:
Fixed support for alternative translation engines.

piotrekbator committed:
Merge branch 'master' of https://github.com/srsgores/ember-form-builder into pr/6

piotrekbator committed:
Fixed unmerged fragments

piotrekbator committed:
Removed unnecessary translation service

piotrekbator commented:
@srsgores fixed and merged. Please, give it a try.

I still need to test the master branch with some real-world projects before releasing a new version.

srsgores commented:
We still need to bump version and push to npm... @piotrekbator

srsgores commented:
Ok, so I guess we'll just have to use

> npm install --save-dev git+https://github.com/nibynic/ember-form-builder.git
